### ISSUE TITLE
Avoid redundant execution,  skip node and single node execution.

### ISSUE THREAD

**Issue by eromoe** [elyra-ai/elyra#2864]
**Is your feature request related to a problem? Please describe.**

I am use elyra to develop local pipeline .  I have some data transform completed which no need process in future,  but `run pipeline` still execute it . 

**Describe the solution you'd like**

## Avoid redundant execution: 
1. a node can return a hash key to identify the consistent of itself.
2. if identify doesn't change ,  skip this node

## Manual set node to `Skip` state
1. mark a node as skip state, so pipeline would not run this node
2. we usually would add some visulization nodes, which we don't want run it all the time, which slow down the whole execution.

## Single node execution
1. select  a node
2. click single execution buttion
3. only execute this node,  no parent nor children.

**Comment by lresende** [comment]
Is this a dup of #1174

**Comment by lresende** [comment]
> a node can precalculate a id hash key to identify the consistent of itself.
> if id hash key doesn't change , skip this node

How to do that when there is external data involved?

**Comment by eromoe** [comment]
@lresende A pipeline develop know his data.

# Cases
- In local development, usually we have some invariable datasets ( or it changes in low   frequency , maybe some days or weeks)
- In online system, we would do some retries ,  but we know input datasets don't change in some range , for example 1 hour or 1 day.

# Design the hash key
depend on the cases , we can design out hash key 

Fixed hash
- formated_time + path

Dependency hash
- formated_time + parent_node_hash

# So that we can avoid redundant calculation

**Comment by eromoe** [comment]
Only the `3rd` point is some extent dup of https://github.com/elyra-ai/elyra/issues/1174 ,  can be merge to 2 button
1. run single without  children nodes
2. run single with children nodes (run downstream)

**Comment by eromoe** [comment]
@lresende Hello? I saw the tag doesn't change, this is not a duplicated issue .

**Comment by kevin-bates** [comment]
I agree with @lresende and his response here: https://github.com/elyra-ai/elyra/issues/1174#issuecomment-759610127

> 1. run single without children nodes

This is possible today via the "submit as pipeline" button - as Luciano points out in the previous link

> 2. run single with children nodes (run downstream)

This is the intent of #1174.

**Comment by lresende** [comment]
We can generalize to a certain point as @kevin-bates mentioned above, but any additional magic will always not be 100% due to scenarios where external dependencies/data are being processed.
